### PR TITLE
Bookstore: Create and Remove Books

### PR DESCRIPTION
**Pull request by DanSam5K**
## In this milestone I accomplished the following:
- I configure the Redux Store (/src/redux/configureStore.js) and importing the necessary methods from Redux:
  - [x] I Import createStore
  - [x] Import combineReducers 
  - [x] Import applyMiddleware
- I write my book's actions and reducer using the [ducks pattern](https://github.com/erikras/ducks-modular-redux).
- In my React component responsible for adding new books set the data inputs in the local React state
- In my React component responsible for removing books
- I use the redux-logger to check if my application is working correctly - by opening my browser's console and look for the logger messages
   - [x] prev state
   - [x] action
   - [x] next state
 - Fix all linters errors

### CODE REVIEW

File: src/components/Books.js [hunk 1]
```diff
@@ -1,50 +1,14 @@
...
-      ],
-    };
-  }
-
-  render() {
-    const { books } = this.state;
-    return (
-      <div>
-        <ShowBooks bookList={books} />
-        <AddNewBook id="categories" name="categories" selected />
```
Comment (henatan99) [inline src/components/Books.js:44]: - [x] Your `AddNewBook` has a callback prop `addBook` which is expected to be passed from this parent component. But there is no prop passed here. As a result, the AddNewBook functionality is not working at all. Kindly fix this issue. :pray: ALso, please don't forget that the `addBook` call back prop is supposed to be a redux action. 

image

File: src/redux/books/bookReducer.js [hunk 2]
```diff
@@ -1,13 +1,48 @@
...
+    author: 'author',
+    genre: 'Economy',
+  },
+  {
+    id: uuidv4(),
+    title: 'Book Three',
+    author: 'author',
+    genre: 'Documentation',
+  },
+];
```
Comment (henatan99) [inline src/redux/books/bookReducer.js]: - [x] As per the project's requirement the initial state of the books object is an empty array. Kindly make changes here. :pray:

File: src/redux/books/bookReducer.js [hunk 3]
```diff
@@ -1,13 +1,48 @@
-import * as Actions from './bookAction';
+import { v4 as uuidv4 } from 'uuid';
+import * as actions from '../../action/bookAction';
```
Comment (henatan99) [inline src/redux/books/bookReducer.js]: - [x] You should assign values for the action types here as per the [ducks pattern](https://github.com/erikras/ducks-modular-redux) indicated in the project's requirement. Kindly make changes here. :pray:

File: src/redux/books/bookReducer.js [hunk 4]
```diff
@@ -1,13 +1,48 @@
...
+    genre: 'Documentation',
+  },
+];
+
+const books = (state = initialBooks, action) => {
   switch (action.type) {
-    case Actions.ADDED_BOOK:
-    case Actions.REMOVED_BOOK:
-    case Actions.SHOWED_BOOK:
+    case actions.ADDED_BOOK: {
```
Comment (henatan99) [inline src/redux/books/bookReducer.js]: - [x] You did not define actions yet. Actions are defined with type and payload attributes. Payload is optional and it is passed as an argument in the action method. Please see the below general example of how to simply define a redux action. Kindly make changes here. :pray: In that case, you don't need to say `case actions.ADDED_BOOK` but instead you can say `case: ADDED_BOOK`. In addition, the action type naming is better if you say `ADD_BOOK` and `REMOVE_BOOK` as it is an action and not a past event. 

```
export const action = (payload) => ({
  type: ACTION_TYPE,
  payload,
});
```

**Review by henatan99 — CHANGES_REQUESTED** [review]
Hi @DanSam5K , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>


## Required Changes ♻️
- [x] I don't recommend using store `<Provider />` in your  `Books` component as you need the store for your `Categories` component too, later. It is always good to use store `<Provider />` in top-level components. :pray: 

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/components/Books.js (hunk 1 above)
Comment (DanSam5K) [reply src/components/Books.js:44]: Noted

File: src/redux/books/bookReducer.js (hunk 2 above)
Comment (DanSam5K) [reply src/redux/books/bookReducer.js]: - Alright Noted

File: src/redux/books/bookReducer.js (hunk 3 above)
Comment (DanSam5K) [reply src/redux/books/bookReducer.js]: - wasn't clear at first what I needed to do, but look into that thank you!

File: src/redux/books/bookReducer.js (hunk 4 above)
Comment (DanSam5K) [reply src/redux/books/bookReducer.js]: Noted

File: src/redux/books/bookReducer.js (hunk 2 above)
Comment (henatan99) [reply src/redux/books/bookReducer.js]: This issue is well addressed! :+1:

File: src/redux/books/bookReducer.js (hunk 3 above)
Comment (henatan99) [reply src/redux/books/bookReducer.js]: Thank you for making the changes and sorry I didn't see your separate `actions` folder which was supposed to be inside the `redux` directory. Anyways your new actions and reducers are far cleaner. :clap:

File: src/redux/books/bookReducer.js (hunk 4 above)
Comment (henatan99) [reply src/redux/books/bookReducer.js]: Sorry once again that I did not see your `actions` folder which in fact was supposed to be in the `redux` directory while you kept it in the root directory. Now you have made the right changes and I think you don't need the `actions` folder in the root directory. :clap:

File: src/components/Books.js (hunk 1 above)
Comment (henatan99) [reply src/components/Books.js:44]: - [x] Unfortunately, you did not make changes on this one. And the `addNewBook` feature is not working on your app.  You should pass the `addBook` callback prop in the `AddNewBook` component. Also please don't forget to use the `useDispath()` hook as the `addBook` method is a redux action method. :pray: 

image

**Review by henatan99 — CHANGES_REQUESTED** [review]
Hi @DanSam5K , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by DanSam5K** [comment]
- I will try to do that sir. But I am kind of shot of ideas. Thank you by the way for pointing it out 👍 
Please if you won't mind helping me see clearly if you have the time? I will appreciate sir.

**Comment by henatan99** [comment]
> * I will try to do that sir. But I am kind of shot of ideas. Thank you by the way for pointing it out +1
>   Please if you won't mind helping me see clearly if you have the time? I will appreciate sir.

Sure I can help you.

**Comment by DanSam5K** [comment]
> > * I will try to do that sir. But I am kind of shot of ideas. Thank you by the way for pointing it out +1
> >   Please if you won't mind helping me see clearly if you have the time? I will appreciate sir.
> 
> Sure I can help you.

Alright I have scheduled a zoom session on slack in regards 🤝 Thanks

File: src/redux/books/bookReducer.js (hunk 4 above)
Comment (DanSam5K) [reply src/redux/books/bookReducer.js]: No problem

File: src/components/Form.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+              <option>Action</option>
+              <option>Drama</option>
+              <option>Romance</option>
+            </Form.Select>
+          </Form.Group>
+          <Button as={Col} variant="primary" type="submit" onClick={addBook}>
+            ADD BOOK
+          </Button>
+        </Row>
+      </Form> */}
```
Comment (EricMbouwe) [inline src/components/Form.js]: - [x] Please consider removing these comments from your code, they make your code look messy.

**Review by EricMbouwe — APPROVED** [review]
Hi @DanSam5K, 

# Status: Approved✔️
Excellent job on this project. Keep it up 👏
Your project is complete! There is nothing else to say other than... it's time to merge it 💪
Congratulations! 🎉

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/components/Form.js (hunk 5 above)
Comment (DanSam5K) [reply src/components/Form.js]: - Thank you noted, forgot to take them out

**Pull request merged**